src/main/java/com/puppycrawl/tools/checkstyle/filters/XpathFilter.java

```diff
@@ -0,0 +1,184 @@
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.filters;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+import com.puppycrawl.tools.checkstyle.TreeWalkerAuditEvent;
+import com.puppycrawl.tools.checkstyle.TreeWalkerFilter;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtils;
+import com.puppycrawl.tools.checkstyle.xpath.AbstractNode;
+import com.puppycrawl.tools.checkstyle.xpath.RootNode;
+import net.sf.saxon.om.Item;
+import net.sf.saxon.sxpath.XPathDynamicContext;
+import net.sf.saxon.sxpath.XPathEvaluator;
+import net.sf.saxon.sxpath.XPathExpression;
+import net.sf.saxon.trans.XPathException;
+
+/**
+ * This filter processes {@link TreeWalkerAuditEvent}
+ * objects based on the criteria of file, check, module id, xpathQuery.
+ *
+ * @author Timur Tibeyev.
+ */
+public class XpathFilter implements TreeWalkerFilter {
+    /** The regexp to match file names against. */
+    private final Pattern fileRegexp;
+
+    /** The pattern for file names. */
+    private final String filePattern;
+
+    /** The regexp to match check names against. */
+    private final Pattern checkRegexp;
+
+    /** The pattern for check class names. */
+    private final String checkPattern;
+
+    /** Module id filter. */
+    private final String moduleId;
+
+    /** Xpath expression. */
+    private final XPathExpression xpathExpression;
+
+    /** Xpath query. */
+    private final String xpathQuery;
+
+    /**
+     * Creates a {@code XpathElement} instance.
+     * @param files regular expression for names of filtered files
+     * @param checks regular expression for filtered check classes
+     * @param moduleId the module id
+     * @param query the xpath query
+     */
+    public XpathFilter(String files, String checks,
+                       String moduleId, String query) {
+        filePattern = files;
+        fileRegexp = Pattern.compile(files);
+        checkPattern = checks;
+        if (checks == null) {
+            checkRegexp = null;
+        }
+        else {
+            checkRegexp = CommonUtils.createPattern(checks);
+        }
+        this.moduleId = moduleId;
+        xpathQuery = query;
+        if (xpathQuery == null) {
+            xpathExpression = null;
+        }
+        else {
+            final XPathEvaluator xpathEvaluator = new XPathEvaluator();
+            try {
+                xpathExpression = xpathEvaluator.createExpression(xpathQuery);
+            }
+            catch (XPathException ex) {
+                throw new IllegalStateException("Unexpected xpath query", ex);
+            }
+        }
+    }
+
+    @Override
+    public boolean accept(TreeWalkerAuditEvent event) {
+        return !isFileNameAndModuleAndCheckNameMatching(event)
+                || !isXpathQueryMatching(event);
```

Comment:
@rnveach 

> So shouldn't your method be like:
> !isFileNameAndModuleAndCheckNameMatching(event) && !isXpathQueryMatching(event);

No, image situation when `isFileNameAndModuleAndCheckNameMatching(event)` returns true and TreeWalker audit event does not match xpath query, so isXpathQueryMatching(event) is false, result expression will be FALSE AND TRUE which is equivalent to FALSE, so message will be removed and event would be suppressed, it is not true

In my implementation `accept` will return false only if `isFileNameAndModuleAndCheckNameMatching` is true  and `isXpathQueryMatching` is true, otherwise method will return true, which is logically correct I think

Maybe rewriting expression to the following form will make it more understandable
`!(isFileNameAndModuleAndCheckNameMatching(event) && isXpathQueryMatching(event));`